pkg/beacon/client/multi.go

```diff
@@ -0,0 +1,202 @@
+
+	// Check each beacon-node sync status
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	for _, instance := range b.Clients {
+		wg.Add(1)
+		go func(client BeaconNode) {
+			defer wg.Done()
+			log := b.Log.WithField("endpoint", client.Endpoint())
+
+			syncStatus, err := client.SyncStatus()
+			if err != nil {
+				log.WithError(err).Error("failed to get sync status")
+				return
+			}
+
+			mu.Lock()
+			defer mu.Unlock()
+
+			if foundSyncedNode {
+				return
+			}
+
+			if bestSyncStatus == nil {
+				bestSyncStatus = syncStatus
+			}
+
+			if !syncStatus.IsSyncing {
+				bestSyncStatus = syncStatus
+				foundSyncedNode = true
+			}
+		}(instance)
+	}
+
+	// Wait for all requests to complete...
+	wg.Wait()
+
+	if !foundSyncedNode {
+		return nil, ErrBeaconNodeSyncing
+	}
+
+	if bestSyncStatus == nil {
+		return nil, ErrNodesUnavailable
+	}
+
+	return bestSyncStatus, nil
+}
+
+func (b *MultiBeaconClient) KnownValidators(headSlot structs.Slot) (AllValidatorsResponse, error) {
+	// return the first successful beacon node response
+	clients := b.clientsByLastResponse()
+
+	for i, client := range clients {
+		log := b.Log.WithField("endpoint", client.Endpoint())
+
+		validators, err := client.KnownValidators(headSlot)
+		if err != nil {
+			log.WithError(err).Error("failed to fetch validators")
+			continue
+		}
+
+		b.bestBeaconIndex.Store(int64(i))
+
+		// Received successful response. Set this index as last successful beacon node
+		return validators, nil
+	}
+
+	return AllValidatorsResponse{}, ErrNodesUnavailable
+}
+
+func (b *MultiBeaconClient) Genesis() (genesisInfo structs.GenesisInfo, err error) {
+	clients := b.clientsByLastResponse()
+	for _, client := range clients {
+		if genesisInfo, err = client.Genesis(); err != nil {
+			b.Log.WithError(err).
+				WithField("endpoint", client.Endpoint()).
+				Warn("failed to get genesis info")
+			continue
+		}
+
+		return genesisInfo, nil
+	}
+
+	return genesisInfo, err
+}
+
+func (b *MultiBeaconClient) Endpoint() string {
+	return b.clientsByLastResponse()[0].Endpoint()
+}
+
+// beaconInstancesByLastResponse returns a list of beacon clients that has the client
+// with the last successful response as the first element of the slice
+func (b *MultiBeaconClient) clientsByLastResponse() []BeaconNode {
+	index := b.bestBeaconIndex.Load()
+	if index == 0 {
+		return b.Clients
+	}
+
+	instances := make([]BeaconNode, len(b.Clients))
+	copy(instances, b.Clients)
+	instances[0], instances[index] = instances[index], instances[0]
+
+	return instances
+}
+
+func (b *MultiBeaconClient) PublishBlock(block *types.SignedBeaconBlock) (err error) {
+	for _, client := range b.clientsByLastResponse() {
+		if err = client.PublishBlock(block); err != nil {
+			b.Log.WithError(err).
+				WithField("endpoint", client.Endpoint()).
+				Warn("failed to publish block to beacon")
+			continue
+		}
+
+		return nil
+	}
+
+	return err
+}
+
+func (b *MultiBeaconClient) AttachMetrics(m *metrics.Metrics) {
```

Comment:
It would be better for Clients to take care of it themselves wherever they are initialized. We don't need this in an interface as well.